# Статья-исследование "Развитие игровой интустрии с 2000 по 2013 год"

- Автор: Сереженко Елена

### Цели и задачи проекта

Команда игры «Секреты Темнолесья» хочет привлечь новую аудиторию и подготовить статью о развитии индустрии игр в начале XXI века. В статье-исследовании необходимо сделать обзор игровых платформ, изучить объёмы продаж игр разных жанров и региональные предпочтения игроков. 

Основные задачи:
* комплексный анализ развития игровой индустрии;
* анализ объёмов продаж игр разных жанров и региональные предпочтения игроков.


Перед началом проведения анализа необходимо дополнительно сделать следующее:

* отобрать данные по времени выхода игры (с 2000 по 2013 год включительно);
* категоризовать игры по оценкам пользователей и экспертов, и выделить три категории:
 * **высокая оценка** — с оценкой от 8 до 10 и от 80 до 100, включая правые границы интервалов;
 * **средняя оценка** — с оценкой от 3 до 8 и от 30 до 80, не включая правые границы интервалов;
 * **низкая оценка** — с оценкой от 0 до 3 и от 0 до 30, не включая правые границы интервалов;
* выделить топ-7 платформ по количеству игр, выпущенных за весь требуемый период.

### Описание данных

Данные `/datasets/new_games.csv` собраны коллегами из открытых источников и содержат информацию о продажах игр разных жанров и платформ, а также пользовательские и экспертные оценки игр:

* `Name` — название игры;
* `Platform` — название платформы;
* `Year of Release` — год выпуска игры;
* `Genre` — жанр игры;
* `NA sales` — продажи в Северной Америке (в миллионах проданных копий);
* `EU sales` — продажи в Европе (в миллионах проданных копий);
* `JP sales` — продажи в Японии (в миллионах проданных копий);
* `Other sales` — продажи в других странах (в миллионах проданных копий);
* `Critic Score` — оценка критиков (от 0 до 100);
* `User Score` — оценка пользователей (от 0 до 10);
* `Rating` — рейтинг организации ESRB (англ. Entertainment Software Rating Board). Эта ассоциация определяет рейтинг компьютерных игр и присваивает им подходящую возрастную категорию.

## Содержание проекта
   
* [1. Загрузка данных и знакомство с ними.](#1.-Загрузка-данных-и-знакомство-с-ними)
* [2. Проверка ошибок в данных и их предобработка.](#2.-Проверка-ошибок-в-данных-и-их-предобработка)
   * [2.1. Названия столбцов датафрейма.](#21-Названия-столбцов-датафрейма)
   * [2.2. Типы данных.](#22-Типы-данных)
   * [2.3. Наличие пропусков в данных.](#23-Наличие-пропусков-в-данных)
   * [2.4. Явные и неявные дубликаты.](#24-Явные-и-неявные-дубликаты)
* [3. Фильтрация данных.](#3.-Фильтрация-данных)
* [4. Категоризация данных.](#4.-Категоризация-данных)
* [5. Итоговый вывод.](#5.-Итоговый-вывод)

### 1. Загрузка данных и знакомство с ними

In [1]:
# Импортируем библиотеку pandas
import pandas as pd

In [2]:
# Выгрузим данные в Датафрейм
df = pd.read_csv('https://code.s3.yandex.net/datasets/new_games.csv')

In [3]:
# Выведем информацию о Датафрейме
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Name             16954 non-null  object 
 1   Platform         16956 non-null  object 
 2   Year of Release  16681 non-null  float64
 3   Genre            16954 non-null  object 
 4   NA sales         16956 non-null  float64
 5   EU sales         16956 non-null  object 
 6   JP sales         16956 non-null  object 
 7   Other sales      16956 non-null  float64
 8   Critic Score     8242 non-null   float64
 9   User Score       10152 non-null  object 
 10  Rating           10085 non-null  object 
dtypes: float64(4), object(7)
memory usage: 1.4+ MB


#### Общий анализ датасета:

* Датасет `/datasets/new_games.csv` содержит 16956 строк и 11 столбцов, в которых представлена информация об играх;
* 4 столбца содержат тип данных **float64**(числа с плавающей). 7 столбцов содержат тип данных **object**(строковые);
* 6 из 11 столбцов содержат пропуски: `Name`, `Year of Release`, `Genre`, `Critic Score`, `User Score`, `Rating`;
* Необходимо преобразовать названия столбцов в `snake case`.

In [4]:
# Выведем первые строки Датафрейма на экран
df.head()

,Name,Platform,Year of Release,Genre,NA sales,EU sales,JP sales,Other sales,Critic Score,User Score,Rating
0,Wii Sports,Wii,2006.0,Sports,41.36,28.96,3.77,8.45,76.0,8,E
1,Super Mario Bros.,NES,1985.0,Platform,29.08,3.58,6.81,0.77,NaN,NaN,NaN
2,Mario Kart Wii,Wii,2008.0,Racing,15.68,12.76,3.79,3.29,82.0,8.3,E
3,Wii Sports Resort,Wii,2009.0,Sports,15.61,10.93,3.28,2.95,80.0,8,E
4,Pokemon Red/Pokemon Blue,GB,1996.0,Role-Playing,11.27,8.89,10.22,1.00,NaN,NaN,NaN


#### Типы данных и их корректность:

* **float64**(числовые значения с плавающей точкой):
 * столбцы `NA sales`, `Other sales` имеют корректный тип данных, для уменьшения объёма используемой памяти можно понизить разрядность;
 * столбец `Critic Score` приводим к **int8**, так как оценка критиков представлена целыми числами от **0** до **100**;
 * для столбца `Year of Release` тип **float64** некорректен. Для работы со временем лучше использовать тип данных **datetime64**. Нам необходим только год, но через **.dt.year** присвоится тип данных отличный от даты. Наиболее подходящим будет тип **int16**;
* **object**(строковые данные):
 * столбцы `Name`, `Platform`, `Genre`, `Rating` содержат название игры, название платформы, жанр игры и нечисловой рейтинг организации ESRB соответственно и задаются подходящим строковым типом данных **object**;
 * столбцы `EU sales`, `JP sales`, `User Score` содержат числовые значения с плавающей точкой, а тип определён как **object**. Необходимо преобразовать данные колонки в тип данных **float32**.

In [5]:
# Изучим пропуски в необходимых для группировки столбцах с рейтингом:
# оценка критиков
df['Critic Score'].unique()

array([76., nan, 82., 80., 89., 58., 87., 91., 61., 97., 95., 77., 88.,
       83., 94., 93., 85., 86., 98., 96., 90., 84., 73., 74., 78., 92.,
       71., 72., 68., 62., 49., 67., 81., 66., 56., 79., 70., 59., 64.,
       75., 60., 63., 69., 50., 25., 42., 44., 55., 48., 57., 29., 47.,
       65., 54., 20., 53., 37., 38., 33., 52., 30., 32., 43., 45., 51.,
       40., 46., 39., 34., 35., 41., 36., 28., 31., 27., 26., 19., 23.,
       24., 21., 17., 22., 13.])

In [6]:
# оценка пользователей
df['User Score'].unique()

array(['8', nan, '8.3', '8.5', '6.6', '8.4', '8.6', '7.7', '6.3', '7.4',
       '8.2', '9', '7.9', '8.1', '8.7', '7.1', '3.4', '5.3', '4.8', '3.2',
       '8.9', '6.4', '7.8', '7.5', '2.6', '7.2', '9.2', '7', '7.3', '4.3',
       '7.6', '5.7', '5', '9.1', '6.5', 'tbd', '8.8', '6.9', '9.4', '6.8',
       '6.1', '6.7', '5.4', '4', '4.9', '4.5', '9.3', '6.2', '4.2', '6',
       '3.7', '4.1', '5.8', '5.6', '5.5', '4.4', '4.6', '5.9', '3.9',
       '3.1', '2.9', '5.2', '3.3', '4.7', '5.1', '3.5', '2.5', '1.9', '3',
       '2.7', '2.2', '2', '9.5', '2.1', '3.6', '2.8', '1.8', '3.8', '0',
       '1.6', '9.6', '2.4', '1.7', '1.1', '0.3', '1.5', '0.7', '1.2',
       '2.3', '0.5', '1.3', '0.2', '0.6', '1.4', '0.9', '1', '9.7'],
      dtype=object)

В столбцах `Critic Score` и `User Score` присутутсвуют значения **NaN**, которые можно интерпретировать как отсутствие рейтинга у игры. В столбце `User Score`есть значения **tbd**, которые означают, что информация ещё не известна или подлежит уточнению.

In [7]:
# Изучим пропуски в необходимых для группировки столбцах с продажами:
# продажи в Европе
df['EU sales'].unique()

array(['28.96', '3.58', '12.76', '10.93', '8.89', '2.26', '9.14', '9.18',
       '6.94', '0.63', '10.95', '7.47', '6.18', '8.03', '4.89', '8.49',
       '9.09', '0.4', '3.75', '9.2', '4.46', '2.71', '3.44', '5.14',
       '5.49', '3.9', '5.35', '3.17', '5.09', '4.24', '5.04', '5.86',
       '3.68', '4.19', '5.73', '3.59', '4.51', '2.55', '4.02', '4.37',
       '6.31', '3.45', '2.81', '2.85', '3.49', '0.01', '3.35', '2.04',
       '3.07', '3.87', '3.0', '4.82', '3.64', '2.15', '3.69', '2.65',
       '2.56', '3.11', '3.14', '1.94', '1.95', '2.47', '2.28', '3.42',
       '3.63', '2.36', '1.71', '1.85', '2.79', '1.24', '6.12', '1.53',
       '3.47', '2.24', '5.01', '2.01', '1.72', '2.07', '6.42', '3.86',
       '0.45', '3.48', '1.89', '5.75', '2.17', '1.37', '2.35', '1.18',
       '2.11', '1.88', '2.83', '2.99', '2.89', '3.27', '2.22', '2.14',
       '1.45', '1.75', '1.04', '1.77', '3.02', '2.75', '2.16', '1.9',
       '2.59', '2.2', '4.3', '0.93', '2.53', '2.52', '1.79', '1.3', '2.6',
   

In [8]:
# продажи в Японии
df['JP sales'].unique()

array(['3.77', '6.81', '3.79', '3.28', '10.22', '4.22', '6.5', '2.93',
       '4.7', '0.28', '1.93', '4.13', '7.2', '3.6', '0.24', '2.53',
       '0.98', '0.41', '3.54', '4.16', '6.04', '4.18', '3.84', '0.06',
       '0.47', '5.38', '5.32', '5.65', '1.87', '0.13', '3.12', '0.36',
       '0.11', '4.35', '0.65', '0.07', '0.08', '0.49', '0.3', '2.66',
       '2.69', '0.48', '0.38', '5.33', '1.91', '3.96', '3.1', '1.1',
       '1.2', '0.14', '2.54', '2.14', '0.81', '2.12', '0.44', '3.15',
       '1.25', '0.04', '0.0', '2.47', '2.23', '1.69', '0.01', '3.0',
       '0.02', '4.39', '1.98', '0.1', '3.81', '0.05', '2.49', '1.58',
       '3.14', '2.73', '0.66', '0.22', '3.63', '1.45', '1.31', '2.43',
       '0.7', '0.35', '1.4', '0.6', '2.26', '1.42', '1.28', '1.39',
       '0.87', '0.17', '0.94', '0.19', '0.21', '1.6', '0.16', '1.03',
       '0.25', '2.06', '1.49', '1.29', '0.09', '2.87', '0.03', '0.78',
       '0.83', '2.33', '2.02', '1.36', '1.81', '1.97', '0.91', '0.99',
       '0.95', '2.0'

Cтолбцы `EU sales`, `JP sales` содержат строковые значения **unknown**.

In [9]:
# Сделаем проверку уникальных значений в жанрах игры
df['Genre'].unique()

array(['Sports', 'Platform', 'Racing', 'Role-Playing', 'Puzzle', 'Misc',
       'Shooter', 'Simulation', 'Action', 'Fighting', 'Adventure',
       'Strategy', nan, 'MISC', 'ROLE-PLAYING', 'RACING', 'ACTION',
       'SHOOTER', 'FIGHTING', 'SPORTS', 'PLATFORM', 'ADVENTURE',
       'SIMULATION', 'PUZZLE', 'STRATEGY'], dtype=object)

В датафрейме присутствуют явные и неявные дубликаты. Пример: `Shooter`, `SHOOTER` и `Simulation`, `SIMULATION`.

In [10]:
# Определим годы, за которые представлены данные в Датафрейме
min_date = df['Year of Release'].min()
max_date = df['Year of Release'].max()
f"Минимум - {min_date}. Максимум - {max_date}"

'Минимум - 1980.0. Максимум - 2016.0'

То есть данные представлены за период с 1980 года по 2016 год.

### 2. Проверка ошибок в данных и их предобработка

 #### 2.1. Названия столбцов датафрейма

In [11]:
# Выведем названия всех столбцов Датафрейма и проверим их стиль написания
df.columns

Index(['Name', 'Platform', 'Year of Release', 'Genre', 'NA sales', 'EU sales',
       'JP sales', 'Other sales', 'Critic Score', 'User Score', 'Rating'],
      dtype='object')

In [12]:
# Приведем названия всех столбцов к стилю snake case
df.columns = df.columns.str.lower().str.replace(' ', '_')
df.columns

Index(['name', 'platform', 'year_of_release', 'genre', 'na_sales', 'eu_sales',
       'jp_sales', 'other_sales', 'critic_score', 'user_score', 'rating'],
      dtype='object')

 #### 2.2. Типы данных

In [13]:
# Уменьшим разрядность:
# преобразуем столбцы через pd.to_numeric для обработки ошибок и выбора разрядности
columns_name = ['na_sales', 'eu_sales', 'jp_sales', 'other_sales']
for column in columns_name:
    df[column] = pd.to_numeric(df[column], errors='coerce', downcast='float')

In [14]:
# Столбец critic_score приводим к int8
# Пропускам присвоим значение-индикатор = -1.
df['critic_score'] = df['critic_score'].fillna(-1).astype('int8')

In [15]:
# Преобразуем year_of_release тип к int16
# Пропускам присвоим значение-индикатор =-1
df['year_of_release'] = df['year_of_release'].fillna(-1).astype('int16')

In [16]:
# Столбец user_score приводим к float
# Присвоим содержащимся в нем строковым значениям tbd пропуски:
df['user_score'] = pd.to_numeric(df['user_score'], errors='coerce').fillna(-1)

In [17]:
# Выведем общую информацию о Датасете после проведения всех преобразований:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16956 entries, 0 to 16955
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   name             16954 non-null  object 
 1   platform         16956 non-null  object 
 2   year_of_release  16956 non-null  int16  
 3   genre            16954 non-null  object 
 4   na_sales         16956 non-null  float32
 5   eu_sales         16950 non-null  float32
 6   jp_sales         16952 non-null  float32
 7   other_sales      16956 non-null  float32
 8   critic_score     16956 non-null  int8   
 9   user_score       16956 non-null  float64
 10  rating           10085 non-null  object 
dtypes: float32(4), float64(1), int16(1), int8(1), object(4)
memory usage: 977.1+ KB


Наш Датафрейм готов к дальнейшему анализу.

 #### 2.3. Наличие пропусков в данных

Посчитать количество пропусков в каждом столбце в абсолютном и относительном значениях.

In [18]:
# Считаем абсолютное количество пропусков
df.isna().sum()

name                  2
platform              0
year_of_release       0
genre                 2
na_sales              0
eu_sales              6
jp_sales              4
other_sales           0
critic_score          0
user_score            0
rating             6871
dtype: int64

In [19]:
# Считаем относительное количество пропусков
round(df.isna().mean()*100, 2)

name                0.01
platform            0.00
year_of_release     0.00
genre               0.01
na_sales            0.00
eu_sales            0.04
jp_sales            0.02
other_sales         0.00
critic_score        0.00
user_score          0.00
rating             40.52
dtype: float64

In [20]:
# В столбцах year_of_release, critic_score пропуски заменены на -1
# Посчитаем: сколько строк содержат -1

year_of_release_sum = len(df[df['year_of_release'] < 0])
critic_score_sum = len(df[df['critic_score'] < 0])
user_score_sum = len(df[df['user_score'] < 0])
year_of_release_per = len(df[df['year_of_release'] < 0]) / len(df) * 100
critic_score_per = len(df[df['critic_score'] < 0]) / len(df) * 100
user_score_per = len(df[df['user_score'] < 0]) / len(df) * 100

print(f" Количесвто пропусков в year_of_release: {year_of_release_sum}")
print(f" Количесвто пропусков в user_score: {user_score_sum}")
print(f" Количесвто пропусков в critic_score: {critic_score_sum}")
print(f" Соотношение пропусков в year_of_release: {round(year_of_release_per,2)}")
print(f" Соотношение пропусков в user_score: {round(user_score_per,2)}")
print(f" Соотношение пропусков в critic_score: {round(critic_score_per,2)}")

 Количесвто пропусков в year_of_release: 275
 Количесвто пропусков в user_score: 9268
 Количесвто пропусков в critic_score: 8714
 Соотношение пропусков в year_of_release: 1.62
 Соотношение пропусков в user_score: 54.66
 Соотношение пропусков в critic_score: 51.39


**Делаем выводы:**

Процент пропусков в столбцах `name`, `genre`, `eu_sales`, `jp_sales` мал, составляет менее 1% данных. Строки с этими пропусками можно удалить.

Проценты пропущенных значений в столбцах `user_score` - 54.66%, `rating` - 40.52%,  `critic_score` - 51.39% составляют почти половину данных в своих же столбцах. Такие данные просто удалить нельзя. 

Выявить точную закономерность не представляется возможным. Можно отметить только тот факт, что в большенстве случаев при отсутствии одной оценки, другие столбцы с оценками будут тоже пусты. 

Можно заменить такие пропуски усреднёнными значениями(рейтинг - мода по жанру игры, оценка - медиана по жанру игры, год предлагаю заменить на год с наибольшим количеством вышедших игр).

In [21]:
# Удалим строки с пропусками в столбцах name, genre, eu_sales, jp_sales
df = df.dropna(subset=['name', 'genre', 'eu_sales', 'jp_sales']) 

In [22]:
# Приведём названия жанров к нижнему регистру
df['genre'] = df['genre'].str.lower()
df['genre'].unique()

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

In [23]:
# Заменим пропуски в столбце rating модой от жанра игры
df['rating'] = df['rating'].fillna(df.groupby('genre')['rating'].transform(lambda x: x.mode()[0]))

In [24]:
# Заменим пропуски в столбце year_of_release модой
# Пропуск = -1
df['year_of_release'] = df['year_of_release'].replace(-1, int(df['year_of_release'].mode()[0]))

In [25]:
# Контрольная проверка на наличие пропусков
df.isna().sum()

name               0
platform           0
year_of_release    0
genre              0
na_sales           0
eu_sales           0
jp_sales           0
other_sales        0
critic_score       0
user_score         0
rating             0
dtype: int64

#### 2.4. Явные и неявные дубликаты в данных

In [26]:
# Проверим наличие неявных дубликатов в столбце platform
df['platform'].unique()

array(['Wii', 'NES', 'GB', 'DS', 'X360', 'PS3', 'PS2', 'SNES', 'GBA',
       'PS4', '3DS', 'N64', 'PS', 'XB', 'PC', '2600', 'PSP', 'XOne',
       'WiiU', 'GC', 'GEN', 'DC', 'PSV', 'SAT', 'SCD', 'WS', 'NG', 'TG16',
       '3DO', 'GG', 'PCFX'], dtype=object)

In [27]:
# Проверим наличие неявных дубликатов в столбце genre
df['genre'].unique()

array(['sports', 'platform', 'racing', 'role-playing', 'puzzle', 'misc',
       'shooter', 'simulation', 'action', 'fighting', 'adventure',
       'strategy'], dtype=object)

In [28]:
# Проверим наличие неявных дубликатов в столбце rating
df['rating'].unique()

array(['E', 'T', 'M', 'E10+', 'K-A', 'AO', 'EC', 'RP'], dtype=object)

Названия платформ, жанров и рейтинга имеют одинаковый стиль написания, нет необходимости в обработке данных столбцов.

In [29]:
# Замена рейтинга 'К-А' на 'Е'
df['rating'] = df['rating'].replace('K-A', 'E')

In [30]:
 for column in ['platform','year_of_release','genre','rating']:
        print(f'Уникальные значения в столбце {column}:')
        print(df[column].sort_values().unique())
        print()

Уникальные значения в столбце platform:
['2600' '3DO' '3DS' 'DC' 'DS' 'GB' 'GBA' 'GC' 'GEN' 'GG' 'N64' 'NES' 'NG'
 'PC' 'PCFX' 'PS' 'PS2' 'PS3' 'PS4' 'PSP' 'PSV' 'SAT' 'SCD' 'SNES' 'TG16'
 'WS' 'Wii' 'WiiU' 'X360' 'XB' 'XOne']

Уникальные значения в столбце year_of_release:
[1980 1981 1982 1983 1984 1985 1986 1987 1988 1989 1990 1991 1992 1993
 1994 1995 1996 1997 1998 1999 2000 2001 2002 2003 2004 2005 2006 2007
 2008 2009 2010 2011 2012 2013 2014 2015 2016]

Уникальные значения в столбце genre:
['action' 'adventure' 'fighting' 'misc' 'platform' 'puzzle' 'racing'
 'role-playing' 'shooter' 'simulation' 'sports' 'strategy']

Уникальные значения в столбце rating:
['AO' 'E' 'E10+' 'EC' 'M' 'RP' 'T']



In [31]:
# Проверим количество явных дубликатов
df.duplicated().sum()

np.int64(241)

В Датафрейме хранится 241 дубликат. Есть смысл удалить эти строки.

In [32]:
new_df = df.drop_duplicates() 
new_df

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
0,Wii Sports,Wii,2006,sports,41.360001,28.959999,3.77,8.45,76,8.0,E
1,Super Mario Bros.,NES,1985,platform,29.080000,3.580000,6.81,0.77,-1,-1.0,E
2,Mario Kart Wii,Wii,2008,racing,15.680000,12.760000,3.79,3.29,82,8.3,E
3,Wii Sports Resort,Wii,2009,sports,15.610000,10.930000,3.28,2.95,80,8.0,E
4,Pokemon Red/Pokemon Blue,GB,1996,role-playing,11.270000,8.890000,10.22,1.00,-1,-1.0,T
...,...,...,...,...,...,...,...,...,...,...,...
16951,Samurai Warriors: Sanada Maru,PS3,2016,action,0.000000,0.000000,0.01,0.00,-1,-1.0,T
16952,LMA Manager 2007,X360,2006,sports,0.000000,0.010000,0.00,0.00,-1,-1.0,E
16953,Haitaka no Psychedelica,PSV,2016,adventure,0.000000,0.000000,0.01,0.00,-1,-1.0,E
16954,Spirits & Spells,GBA,2003,platform,0.010000,0.000000,0.00,0.00,-1,-1.0,E


In [33]:
# Проверим на наличие неявных дубликатов, которые могли записаться в Датафрейм после каких-то сбоев, из разных источников, но с одними данными и так далее.
new_df = new_df.sort_values(by=new_df.columns.tolist())
new_df[new_df.duplicated(subset=['name', 'platform', 'year_of_release', 'genre'],keep=False)]

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
16465,Madden NFL 13,PS3,2012,sports,0.00,0.01,0.0,0.00,83,5.5,E
606,Madden NFL 13,PS3,2012,sports,2.11,0.22,0.0,0.23,83,5.5,E


Найдены две строки для игры `'Madden NFL 13'` **2012** года для **PS3** в жанре **sports** с разными данными по продажам. Можно объединить данные две строки, заполнив необходимые столбцы.

In [34]:
# Объединим данное значение вручную, чтобы не нагружать сервер ради одного вычисления
new_df.loc[606, 'eu_sales'] += 0.01
new_df = new_df.drop_duplicates(subset=['name', 'platform', 'year_of_release', 'genre'])

In [35]:
# Выведем количество строк и столбцов после всех изменений
new_df.shape

(16702, 11)

После всех выполненных преобразований в датасете осталось 16702 строка.
Изначально датасет содержал 16956 строк.
Было удалено 254 строк с данными, что составляет ~1.3% от всего количества.

**Подготовка данных к дальнейшему анализу завершена:**
* преобразование типов данных;
* обработка пропущенных значений по усредненным значениям;
* удаление явных и неявных дубликатов.

### 3. Фильтрация данных

Коллеги хотят изучить историю продаж игр в начале XXI века, и их интересует период с 2000 по 2013 год включительно. Необходимо отобрать данные по этому показателю.

In [36]:
# Сделаем выборку данных с 2000 по 2013 годы
# Сохраним в новый срез данных new_period_df
new_period_df = new_df[new_df['year_of_release'].between(2000, 2013)]

In [37]:
# Выведем сам Датафрейм
new_period_df

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating
3394,Frozen: Olaf's Quest,3DS,2013,platform,0.27,0.27,0.00,0.05,-1,-1.0,E
3906,Frozen: Olaf's Quest,DS,2013,platform,0.21,0.26,0.00,0.04,-1,-1.0,E
2478,Tales of Xillia 2,PS3,2012,role-playing,0.20,0.12,0.45,0.07,71,7.9,T
8460,.hack//G.U. Vol.1//Rebirth,PS2,2006,role-playing,0.00,0.00,0.17,0.00,-1,-1.0,T
7182,.hack//G.U. Vol.2//Reminisce,PS2,2006,role-playing,0.11,0.09,0.00,0.03,-1,-1.0,T
...,...,...,...,...,...,...,...,...,...,...,...
647,uDraw Studio,Wii,2010,misc,1.65,0.57,0.00,0.20,71,-1.0,E
8397,uDraw Studio: Instant Artist,Wii,2011,misc,0.06,0.09,0.00,0.02,-1,-1.0,E
15836,uDraw Studio: Instant Artist,X360,2011,misc,0.01,0.01,0.00,0.00,54,5.7,E
477,wwe Smackdown vs. Raw 2006,PS2,2009,fighting,1.57,1.02,0.00,0.41,-1,-1.0,T


In [38]:
# Проверим минимальный год в Датасете с фильтрацией
new_period_df['year_of_release'].min()

np.int16(2000)

In [39]:
# Проверим максимальный год в Датасете с фильтрацией
new_period_df['year_of_release'].max()

np.int16(2013)

Выборка по годам произведена верно.

### 4. Категоризация данных

Необходимо категоризовать игры по оценкам пользователей и экспертов:
* **высокая оценка** — с оценкой от 8 до 10 и от 80 до 100, включая правые границы интервалов;
* **средняя оценка** — с оценкой от 3 до 8 и от 30 до 80, не включая правые границы интервалов;
* **низкая оценка** — с оценкой от 0 до 3 и от 0 до 30, не включая правые границы интервалов.

In [40]:
# Создадим копию
new_period_df = new_period_df.copy()

In [41]:
# Сделаем категоризацию на основе оценок пользователей
new_period_df['user_category'] = pd.cut(new_period_df['user_score'], bins=[-1,0, 3, 8, 10], labels=['без оценки','низкая оценка', 'средняя оценка', 'высокая оценка'], right=False)

In [42]:
# Сделаем категоризацию на основе оценок критиков
new_period_df['critic_category'] = pd.cut(new_period_df['critic_score'], bins=[-1,0, 30, 80, 100], labels=['без оценки','низкая оценка', 'средняя оценка', 'высокая оценка'], right=False)

In [43]:
# Выведем таблицу после категоризаций
new_period_df

,name,platform,year_of_release,genre,na_sales,eu_sales,jp_sales,other_sales,critic_score,user_score,rating,user_category,critic_category
3394,Frozen: Olaf's Quest,3DS,2013,platform,0.27,0.27,0.00,0.05,-1,-1.0,E,без оценки,без оценки
3906,Frozen: Olaf's Quest,DS,2013,platform,0.21,0.26,0.00,0.04,-1,-1.0,E,без оценки,без оценки
2478,Tales of Xillia 2,PS3,2012,role-playing,0.20,0.12,0.45,0.07,71,7.9,T,средняя оценка,средняя оценка
8460,.hack//G.U. Vol.1//Rebirth,PS2,2006,role-playing,0.00,0.00,0.17,0.00,-1,-1.0,T,без оценки,без оценки
7182,.hack//G.U. Vol.2//Reminisce,PS2,2006,role-playing,0.11,0.09,0.00,0.03,-1,-1.0,T,без оценки,без оценки
...,...,...,...,...,...,...,...,...,...,...,...,...,...
647,uDraw Studio,Wii,2010,misc,1.65,0.57,0.00,0.20,71,-1.0,E,без оценки,средняя оценка
8397,uDraw Studio: Instant Artist,Wii,2011,misc,0.06,0.09,0.00,0.02,-1,-1.0,E,без оценки,без оценки
15836,uDraw Studio: Instant Artist,X360,2011,misc,0.01,0.01,0.00,0.00,54,5.7,E,средняя оценка,средняя оценка
477,wwe Smackdown vs. Raw 2006,PS2,2009,fighting,1.57,1.02,0.00,0.41,-1,-1.0,T,без оценки,без оценки


Игры разбиты на категории на основе оценок пользователей и критиков. Можно сгруппировать полученные данные для подсчёта количества игр.

In [44]:
#Количество игр по категориям на основе оценок пользователей
new_period_df.groupby('user_category', observed=False)['name'].count()

user_category
без оценки        6438
низкая оценка      119
средняя оценка    4157
высокая оценка    2326
Name: name, dtype: int64

In [45]:
#Количество игр по категориям на основе оценок критиков
new_period_df.groupby('critic_category', observed=False)['name'].count()

critic_category
без оценки        5725
низкая оценка       58
средняя оценка    5539
высокая оценка    1718
Name: name, dtype: int64

В разрезе двух категоризаций большинство игр находится в пункте **средняя оценка**.
Меньше всего игр находится в категории **низкая оценка**(0.91% от общего количества оценок пользователя и 0.45% - от оценок критиков).
**Высокую оценку** игры от критиков получали меньше раз, чем от пользоватей.
**Без оценки** в обеих категоризациях составляют довольно большое число, скорее всего из-за неактуальности платформ и площадок сбора информации этих игр, возможно и региональные распределения сыграли свою роль.

Выделим **ТОП-7** платформ по количеству игр, выпущенных за период с 2000 года по 2013 год.

In [46]:
new_period_df.groupby('platform')['name'].count().sort_values(ascending=False).head(7)

platform
PS2     2160
DS      2147
Wii     1309
PSP     1195
X360    1148
PS3     1111
XB       824
Name: name, dtype: int64

Первых два фаворита:
* PS2 - Sony Playstation 2 (2160 игр);
* Nintendo DS (2147 игр).

### 5. Итоговый вывод

Была проведена подготовка датасета `/datasets/new_games.csv`, где представлена информация об играх, их жанрах, годе выпуска, продажах в различных регионах и рейтингах. Они содержат 11 столбцов и 16 956 строк. 

Во время подготовки были проведены следующие шаги:

* загрузка и знакомство с данными;

* предобработка данных(приведение названий столбцов к стандарту);

* преобразование типов данных:
 * столбцы `na_sales`, `eu_sales`, `jp_sales`, `other_sales`- выбор разрядности;
 * столбец `critic_score` к **int8**;
 * столбец `year_of_release` к **int16**;
 * столбец `user_score` к **float**;

* обработка пропусков:
 * удалены пропуски в столбцах name, genre, eu_sales, jp_sales (маленький процент пропусков);
 * пропуски в user_score, critic_score, rating заполнены медианой/модой по жанру;
 * удалены явные дубликаты (241 строка) и неявные дубликаты объдинены.
 
* фильтрация данных(отобраны игры, выпущенные с 2000 по 2013 год включительно (13039 строк));

* категоризация данных:

 *  добавлены новые столбцы:
    * **user_category** — категория оценки пользователей (низкая, средняя, высокая).
    * **critic_category** — категория оценки критиков (низкая, средняя, высокая).

 * выделены **ТОП-7** платформ по количеству выпущенных игр(PS2, DS, Wii, PSP, X360, PS3, XB).

Итоговый датасет new_period_df готов к дальнейшему анализу командой "Секреты Темнолесья" и включает:

* 13040 игр за период с 2000 года по 2013 год;
* 11 исходных столбцов + 2 новых с категориями по оценкам пользователей и критиков соответственно.